include/deal.II/base/tensor_accessors.h

```diff
@@ -0,0 +1,586 @@
+      T &
+      extract(T &t, const TableIndices<rank> &indices)
+      {
+        return t;
+      }
+    };
+  } /* namespace internal */
+
+
+  /**
+   * Return a reference (const or non-const) to a subobject of a tensorial
+   * object T as described by a TableIndices object @p indices. For
+   * example:
+   * @code
+   *   Tensor<5, dim> tensor;
+   *   TableIndices<5> indices (0, 1, 2, 3, 4);
+   *   TensorAccessors::extract(tensor, indices) = 42;
+   * @endcode
+   * This is equivalent to <code>tensor[0][1][2][3][4] = 42.</code>.
+   *
+   * @tparam T A tensorial object of rank @tparam rank. @tparam T must
+   * provide a local typedef <code>value_type</code> and an index operator
+   * <code>operator[]()</code> that returns a (const or non-const)
+   * reference of <code>value_type</code>. Further, its tensorial rank must
+   * be equal or greater than @tparam rank
+   */
+  template<int rank, typename T>
+  typename ReturnType<rank, T>::value_type &
+  extract(T &t, const TableIndices<rank> &indices)
+  {
+    return internal::ExtractHelper<0, rank>::template extract<T>(t, indices);
+  }
+
+
+  namespace internal
+  {
+    // FIXME: Documentation
+    template <int rank_1, int rank_2, int dim>
+    class Contract
+    {
+    public:
+      template<typename T1, typename T2, typename T3>
+      static void contract(T1 &result, const T2 &left, const T3 &right)
+      {
+        for (unsigned int i = 0; i < dim; ++i)
+          Contract<rank_1 - 1, rank_2, dim>::contract(result[i], left[i], right);
+      }
+    };
+
+    // FIXME: Documentation
+    template <int rank_2, int dim>
+    class Contract<1, rank_2, dim>
+    {
+    public:
+      template<typename T1, typename T2, typename T3>
+      static void contract(T1 &result, const T2 &left, const T3 &right)
+      {
+        for (unsigned int i = 0; i < dim; ++i)
+          Contract<1, rank_2 - 1, dim>::contract(result[i], left, right[i]);
+      }
+    };
+
+    // FIXME: Documentation
+    template <int dim>
+    class Contract<1, 1, dim>
+    {
+    public:
+      template<typename T1, typename T2, typename T3>
+      static void contract(T1 &result, const T2 &left, const T3 &right)
+      {
+        result = T1();
+        for (unsigned int i =0; i < dim; ++i)
+          result += left[i] * right[i];
+      }
+    };
+  } /* namespace internal */
+
+
+  /**
+   * FIXME: Documentation
+   *
+   * This function contracts two tensorial objects @p left and @p right and
+   * stores the result in @p result. The contraction is done over the last
+   * index of both tensorial objects. Calling this function is equivalent
+   * of writing the following low level code:
+   * @code
+   *   for(unsigned int i_0 = 0; i_0 < dim; ++i_0)
+   *     ...
+   *       for(unsigned int i_r = 0; i_r < dim; ++i_r)
+   *         {
+   *           result[i_0]..[i_r] = 0.;
+   *           for(unsigned int j = 0; j < dim; ++j)
+   *             result[i_0]..[i_r] += left[i_0]..[i_l] * right[i_l1]..[i_r]
```

Comment:
I think you actually need `[j]` as an additional last index on `left`, and as an additional first index on `right`.